File: dotnet/src/SemanticKernel.Core/PromptTemplate/AggregatorPromptTemplateFactory.cs
// Copyright (c) Microsoft. All rights reserved.

using System.Diagnostics.CodeAnalysis;

namespace Microsoft.SemanticKernel;

/// <summary>
/// Provides a <see cref="IPromptTemplateFactory"/> which aggregates multiple prompt template factories.
/// </summary>
/// <remarks>
/// Attempts via <see cref="TryCreate"/> to create an <see cref="IPromptTemplate"/> from a
/// <see cref="PromptTemplateConfig"/> will iterate through the aggregated factories, using
/// the result from the first to successfully handle the supplied configuration.
/// </remarks>
public sealed class AggregatorPromptTemplateFactory : IPromptTemplateFactory
{
    private readonly IPromptTemplateFactory?[] _promptTemplateFactories;

    /// <summary>Initializes the instance.</summary>
    /// <param name="promptTemplateFactories">Ordered <see cref="IPromptTemplateFactory"/> instances to aggregate.</param>
    public AggregatorPromptTemplateFactory(params IPromptTemplateFactory[] promptTemplateFactories)
    {
        Verify.NotNullOrEmpty(promptTemplateFactories);
        foreach (IPromptTemplateFactory promptTemplateFactory in promptTemplateFactories)
        {
            Verify.NotNull(promptTemplateFactory, nameof(promptTemplateFactories));
        }

        this._promptTemplateFactories = promptTemplateFactories;
    }

    /// <inheritdoc/>
    public bool TryCreate(PromptTemplateConfig templateConfig, [NotNullWhen(true)] out IPromptTemplate? result)
    {
        Verify.NotNull(templateConfig);

        foreach (var promptTemplateFactory in this._promptTemplateFactories)
        {
            if (promptTemplateFactory?.TryCreate(templateConfig, out result) is true && result is not null)
            {
                return true;
            }
        }

        result = null;
        return false;
    }
}

File: java/semantickernel-api/src/main/java/com/microsoft/semantickernel/semanticfunctions/AggregatorPromptTemplateFactory.java
// Copyright (c) Microsoft. All rights reserved.
package com.microsoft.semantickernel.semanticfunctions;

import java.util.ArrayList;
import java.util.List;
import reactor.util.annotation.NonNull;

/**
 * An collection of {@link PromptTemplateFactory} instances. The factory will try to create a
 * {@link PromptTemplate} using each factory in the collection until one is successful.
 */
public class AggregatorPromptTemplateFactory implements PromptTemplateFactory {

    private final List<PromptTemplateFactory> templateFactories;

    /**
     * Creates a new instance of {@link AggregatorPromptTemplateFactory}.
     *
     * @param templateFactories the factories to aggregate
     */
    public AggregatorPromptTemplateFactory(List<PromptTemplateFactory> templateFactories) {
        this.templateFactories = new ArrayList<>(templateFactories);
    }

    @Override
    public PromptTemplate tryCreate(@NonNull PromptTemplateConfig templateConfig) {
        for (PromptTemplateFactory factory : templateFactories) {
            try {
                return factory.tryCreate(templateConfig);
            } catch (UnknownTemplateFormatException ignored) {
                // No-op
            }
        }

        throw new UnknownTemplateFormatException(templateConfig.getTemplateFormat());
    }
}
